test/integration/transaction/tracer.test.js

```diff
@@ -0,0 +1,877 @@
+  const tracer = helper.getTracer()
+  ctx.nr = {
+    agent,
+    tracer
+  }
+})
+
+test.afterEach((ctx) => {
+  helper.unloadAgent(ctx.nr.agent)
+})
+
+test('bind in transaction', async function testBind(t) {
+  const { agent, tracer } = t.nr
+
+  const context = {}
+  const plan = tspl(t, { plan: 10 })
+
+  helper.runInTransaction(agent, function inTrans(transaction) {
+    const root = transaction.trace.root
+    const other = tracer.createSegment('other')
+    plan.equal(tracer.getTransaction(), transaction, 'should start in transaction')
+
+    plan.equal(tracer.getSegment(), root, 'should start at root segment')
+    let bound = tracer.bindFunction(compare)
+
+    tracer.setSegment(null)
+    bound.call(context, root)
+    plan.equal(tracer.getSegment(), null, 'should reset segment after being called')
+
+    bound = tracer.bindFunction(compare, other)
+    bound.call(context, other)
+
+    tracer.setSegment(root)
+    bound = tracer.bindFunction(compare, null)
+
+    plan.equal(tracer.getSegment(), root, 'should be back to root segment')
+    bound.call(context, root)
+
+    function compare(expected) {
+      plan.equal(this, context, 'should pass through context')
+      plan.equal(tracer.getSegment(), expected, 'should have expected segment')
+    }
+  })
+
+  await plan.completed
+})
+
+test('bind outside transaction', async function testBind(t) {
+  const { agent, tracer } = t.nr
+
+  let root
+  const plan = tspl(t, { plan: 5 })
+
+  let bound = tracer.bindFunction(compare)
+  compare(null)
+  bound(null)
+
+  helper.runInTransaction(agent, function inTrans(transaction) {
+    root = transaction.trace.root
+    bound(null)
+  })
+
+  bound = tracer.bindFunction(compare, root)
+  bound(root)
+  compare(null)
+
+  function compare(expected) {
+    plan.equal(tracer.getSegment(), expected)
+  }
+
+  await plan.completed
+})
+
+test('bind + throw', async function testThrows(t) {
+  const { agent, tracer } = t.nr
+
+  const error = new Error('oh no!!')
+  const plan = tspl(t, { plan: 12 })
+
+  helper.runInTransaction(agent, function inTrans(transaction) {
+    const root = transaction.trace.root
+    compare(dangerous(root, root), root)
+    compare(dangerous(null, root), root)
+
+    tracer.setSegment(null)
+    compare(dangerous(root, root), null)
+    compare(dangerous(null, null), null)
+  })
+
+  function compare(run, expected) {
+    try {
+      run()
+    } catch (err) {
+      plan.equal(err, error, 'should have expected error')
+      plan.equal(tracer.getSegment(), expected, 'should catch in context')
+    }
+  }
+
+  function dangerous(segment, expected) {
+    return tracer.bindFunction(function bound() {
+      plan.equal(tracer.getSegment(), expected, 'should have expected segment')
+      throw error
+    }, segment)
+  }
+  await plan.completed
+})
+
+test('bind + capture error', async function testThrows(t) {
+  const { agent, tracer } = t.nr
+  const listeners = process.listeners('uncaughtException')
```

Comment:
I'm going to look into using the helper that I think @jsumners-nr wrote for this

Reply:
looks like i can't use it because the handler needs context from a closure above it

Reply:
You would probably use https://github.com/newrelic/node-newrelic/blob/36bc0e7f97f1bbde522031942ec0cc7c1d4c1349/test/lib/temp-override-uncaught.js#L32

Your passed in handler would still retain the closure.

Reply:
I can just remove the listeners with the `temp-remove-listeners`